clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/AbstractGoogleServerGroupCachingAgentTest.java

```diff
@@ -0,0 +1,748 @@
+
+    instanceGroupManager.setTargetPools(null);
+    loadBalancerItem.setValue(null);
+    cacheResult = cachingAgent.loadData(inMemoryProviderCache());
+    serverGroup = getOnlyServerGroup(cacheResult);
+    assertThat(serverGroup.getDisabled()).isFalse();
+  }
+
+  @Test
+  void serverGroupAutoscalerProperties() {
+
+    InstanceGroupManager instanceGroupManager =
+        new InstanceGroupManager().setName("myServerGroup").setTargetSize(303).setZone(ZONE_URL);
+    Autoscaler autoscaler =
+        new Autoscaler()
+            .setZone(ZONE_URL)
+            .setTarget("myServerGroup")
+            .setAutoscalingPolicy(
+                new AutoscalingPolicy().setMinNumReplicas(101).setMaxNumReplicas(202))
+            .setStatusDetails(
+                ImmutableList.of(
+                    new AutoscalerStatusDetails().setMessage("message1"),
+                    new AutoscalerStatusDetails().setMessage("message2")));
+
+    Compute compute =
+        new StubComputeFactory()
+            .setInstanceGroupManagers(instanceGroupManager)
+            .setAutoscalers(autoscaler)
+            .create();
+    AbstractGoogleServerGroupCachingAgent cachingAgent =
+        createCachingAgent(
+            compute, ImmutableList.of(instanceGroupManager), ImmutableList.of(autoscaler));
+
+    // If there are load balancers and target pools, then we not disabled
+    CacheResult cacheResult = cachingAgent.loadData(inMemoryProviderCache());
+    GoogleServerGroup serverGroup = getOnlyServerGroup(cacheResult);
+    assertThat(serverGroup.getAsg())
+        .containsOnly(entry("minSize", 101), entry("maxSize", 202), entry("desiredCapacity", 303));
+    assertThat(serverGroup.getAutoscalingMessages()).containsExactly("message1", "message2");
+  }
+
+  public static AbstractGoogleServerGroupCachingAgent createCachingAgent(
+      Compute compute, Collection<InstanceGroupManager> instanceGroupManagers) {
+    return createCachingAgent(
+        compute, instanceGroupManagers, /* autoscalers= */ ImmutableList.of());
+  }
+
+  public static AbstractGoogleServerGroupCachingAgent createCachingAgent(
+      Compute compute,
+      Collection<InstanceGroupManager> instanceGroupManagers,
+      Collection<Autoscaler> autoscalers) {
+    return new AbstractGoogleServerGroupCachingAgent(
+        new GoogleNamedAccountCredentials.Builder()
+            .project(PROJECT)
+            .name(ACCOUNT_NAME)
+            .compute(compute)
+            .regionToZonesMap(ImmutableMap.of(REGION, ImmutableList.of(ZONE)))
+            .build(),
+        new GoogleComputeApiFactory(
+            new GoogleOperationPoller(),
+            new DefaultRegistry(),
+            "user-agent",
+            MoreExecutors.listeningDecorator(Executors.newCachedThreadPool())),
+        new DefaultRegistry(),
+        REGION,
+        new ObjectMapper()) {
+      @Override
+      Collection<String> getOnDemandKeysToEvictForMissingServerGroup(
```

Comment:
Is there some way to annotate all these methods with `@ParametersAreNonnullByDefault`, as they are overriding methods that have that annotation.  Maybe there's not a clean way, in which case it probably doesn't make sense to clutter up this test code with it.

Reply:
I did `Refactor→Convert Anonymous to Inner`, which is probably better anyway... an anonymous class with 8 method implementations is pretty ugly, thanks.